[MRG+1] Correcting length of explained_variance_ratio_, eigen solver

JPFrancoia opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fix #6032 
#### What does this implement/fix? Explain your changes.

Attribute explained_variance_ratio_ from LinearDiscriminantAnalysis class will be of length n_components (eigen solver).
#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Attribute explained_variance ratio is now available for LDA with the SVD
solver. See issue #5216. However, this attribute is corrected from the
previous commit which, might not return the expected values.

Commit:
Corrected the dimension of the array explained_variance_ratio_ for LDA,
solver svd.

Commit:
Cleaned the calculation of explained_variance_ratio_ in
LinearDiscriminantAnalysis, SVD solver.

Commit:
Regression test for explained_variance_ratio_, LDA, SVD solver.

Used the script of @hlin117 in #5216 to generate better test data,
make_blobs was not good enough.

JPFrancoia committed:
Added 4 spaces to conform to PEP8.

JPFrancoia committed:
Modified the length of the slice: now uses a variable, not a hardcoded
number.

JPFrancoia committed:
Modified the What's new file to mention the bug fix about the wrong
results returned with discriminant_analysis.LinearDiscriminantAnalysis's
attribute explained_variance_ratio.

willingc committed:
Reorganize README

Remove write clone

Update install

Update mailing list url

Fix changes re: @amuller

Add @raghavrv suggested changes

stuppie committed:
use sparse contingency matrix for supervised cluster metrics
Remove max_n_classes option

amueller committed:
clean up deprecation warning stuff in common tests

minor fixes in preprocessing tests

jnothman committed:
ENH raise error when scoring=sklearn.metrics.{classification,ranking,...}.*

themrmax committed:
make export_graphviz return string if out_file is None

remove unused StringIO.get_value

fix lint issues

only return string if out_file is None

fix indentation

fix lint error

retain default behaviour of out_file=tree.dot

add depreciation warning for out_file default

use double quotes

edit depreciation warning and add versionadded to docstring

only show DepreciationWarning if out_file paramater is actually missing

remove lambda for flake8 linting

add object inheritence to class definition

jnothman committed:
TST fix generated test names in case where test fails

ogrisel committed:
Release 0.18rc

amueller committed:
make pickle version test more stable (#7415)

ogrisel committed:
Release 0.18rc1

jnothman committed:
TST move test naming and use inspect

ogrisel committed:
FIX use == to avoid relying on string interning

TomDLT committed:
DOC fix various syntax in whats_new (#7421)

GaelVaroquaux committed:
ENH: display the original exception (#5906)

Fixes #5905

GaelVaroquaux committed:
ENH: display the original exception (#5906)

Fixes #5905

ogrisel committed:
FIX missing import in setup.py

ogrisel committed:
Release 0.18rc2

Commit:
lda_qda doc: QDA vs Naive Bayes correction (#5997)

* lda_qda doc: QDA vs Naive Bayes correction

Small correction in lda_qda.rst re: comparison of QDA with diagonal covariance matrices and Naive Bayes.

In Naive Bayes, it doesn't make sense to talk about conditional independence between the classes. The Naive Bayes assumptions states that given the label, the input features are conditionally independent. This is reflected in the QDA by assuming diagonal covariance matrices for each class.

* Update lda_qda.rst

hallr committed:
added contributors to whats new for 0.16 and 0.17 (#5840)

* added contributors to whats new for 0.16 and 0.17

* added contributors for 016 and 017 as single line, comma delimited

* added contributors for 016 and 017 as single line, comma delimited

amueller committed:
added contributors for 0.18, minor formatting fixes.

s/Contributors/Code Contributors/g

add some more entries to the mailmap

more mailmap and whatsnew additions

kmike committed:
use == instead of 'is' for string comparison (#7422)

nelson-liu committed:
DOC add species distributions dataset to classes.rst (#7430)

lesteve committed:
Change flake8_diff.sh logic for branch in main repo (#7437)

Only commits that were pushed in the branch are tested. They are
provided by the TRAVIS_COMMIT_RANGE environment variable.

rth committed:
[MRG] Text vectorizers memory usage improvement (#7272)

With thanks to @ephes and @aupiff.

jnothman committed:
Make named_check private

hlin117 committed:
[MRG + 1] DOC Default parameter info for KernelPCA (#6035)

* Added default attribute information and more parameter data to KernelPCA docs

* Pulling back changes from rebase

ogrisel committed:
[MRG+2] Faster isotonic rebased (#7444)

stvn66 committed:
fix typo in DBSCAN explanation

stvn66 committed:
remove unnecessary comma

elena-sharova committed:
Provided a fix for issue #7140 - Minor discrepancy in hashing feature
  transformation example.

aabadie committed:
DOC fix deprecation warning in faces decomposition example

aabadie committed:
[MRG+2] DOC: update persistence documentation with Joblib new behavior (#7434)

aabadie committed:
fixing failing faces decomposition example

aabadie committed:
replace RandomizedPCA by new API in remaining documentation

rsmith54 committed:
Add OneVs{One,All}Classifier._pairwise: fix for #7306 (#7350)

Commit:
MAINT remove unneeded import (#7465)

doshyt committed:
[MRG] Add Deprecation Warning to address making "batch" the default learning method in LDA in 0.19 release (#6999)

* update default LDA method to batch

* update tests to match default LDA learning scheme

* DOC trim information on faster learning scheme with LDA

* Add deprecation warning and revert learning_method back to the default

After the discussion, it was decided that a proper deprecation cycle
for the parameter is required. Therefore, in this PR only the warning is
added. The default learning_method will be changed to 'batch' in 0.19.

* Revert back tests to the default LDA

* DOC deprecate in 0.20 instead of 0.19

jnothman committed:
[MRG+1] Fix regression in silhouette_score for clusters of size 1 (#7438)

Commit:
Addressing issue #7468. (#7472)

Removing entry about "indicator features" in `whats_new.rst`, because it was
causing multiple issues and this resulted in being reverted.

jnothman committed:
[MRG+2] FIX adaboost estimators not randomising correctly (#7411)

* FIX adaboost estimators not randomising correctly

(fixes #7408)

FIX ensure nested random_state is set in ensembles

* DOC add what's new

* Only affect *__random_state, not *_random_state for now

* TST More informative assertions for ensemble tests

* More specific testing of different random_states

lesteve committed:
Fix typo in whats_new.rst

hnykda committed:
Sync docstring and definition default argument in kneighbors (#7476)

mlliou112 committed:
DOC: Added Nested Cross Validation Example (#7111)

amueller committed:
no fixed width within bold

jnothman committed:
DOC more what's new subdivision

amueller committed:
[MRG+2] Timing and training score in GridSearchCV (#7325)

* Resolved issue #6894 and #6895:
    Now *SearchCV.results_ includes both timing and training scores.

wrote new test (sklearn/model_selection/test_search.py)
and new doctest (sklearn/model_selection/_search.py)
added a few more lines in the docstring of GridSearchCV and RandomizedSearchCV.
Revised code according to suggestions.
Add a few more lines to test_grid_search_results():
    1. check test_rank_score always >= 1
    2. check all regular scores (test/train_mean/std_score) and timing >= 0
    3. check all regular scores <= 1
Note that timing can be greater than 1 in general, and std of regular scores
always <= 1 because the scores are bounded between 0 and 1.

* ENH/FIX timing and training score.

* ENH separate fit / score times
* Make score_time=0 if errored; Ignore warnings in test
* Cleanup docstrings
* ENH Use helper to store the results
* Move fit time computation to else of try...except...else
* DOC readable sample scores
* COSMIT Add a commnent on why time test is >= 0 instead of > 0
  (Windows time.time precision is not accurate enought to be non-zero
   for trivial fits)

* Convey that times are in seconds

# Conflicts:
#	doc/whats_new.rst

amueller committed:
REL changes for release in 0.18.X branch (#7414)

amueller committed:
DOC Attempt to impose some order on What's New 0.18

amueller committed:
fix rebase messup

amueller committed:
fix scoring function table

amueller committed:
minor fixes to whatsnew

amueller committed:
fix lbfgs rename (#7503)

amueller committed:
minor doc fixes

amueller committed:
[MRG + 1] More versionadded everywhere! (#7403)

* insert versionadded versionchanged directives in docstrings for 0.18

indicate where exception classes were moved from

* moved versionadded in the proper places

GaelVaroquaux committed:
DOC: trivial rst fix to whats_new.rst (#7511)

Fix for proper rendering

jnothman committed:
[MRG] CI build web site from branches (#7498)

* CI build web site from branches

Fixes #7495

* CI Try one regexp instead of list

amueller committed:
0.18 is no longer in development. thanks @mblondel

amueller committed:
add warning about python 2.6 being no longer supported (#7532)

JPFrancoia committed:
Fix issue #6032. Attribute explained_variance_ratio_ from
LinearDiscriminantAnalysis class will be of length n_components.

JPFrancoia committed:
Merge branch '0.18.X' of https://github.com/scikit-learn/scikit-learn into 0.18.X

jnothman commented:
Please add a non-regression test

JPFrancoia commented:
The test already exists but is incorrect:

https://github.com/scikit-learn/scikit-learn/blob/0.18.X/sklearn/tests/test_discriminant_analysis.py#L161-L186

See https://github.com/scikit-learn/scikit-learn/issues/6032#issuecomment-252450243

JPFrancoia committed:
Fixed indention (PEP8).

jnothman commented:
Then surely you propose to fix the test?

JPFrancoia committed:
Fixed the length of explained_variance_ratio_ for the
LinearDiscriminantAnalysis class, eigen solver, once and for good. Fixes

JPFrancoia commented:
Test fixed.

jnothman commented on sklearn/discriminant_analysis.py:
```diff
@@ -330,13 +330,21 @@ class scatter). This solver supports both classification and
         Sb = St - Sw  # between scatter
 
+        # Get maximum length for explained_variance_ratio_
+        if n_classes < self.n_components:
```
can you just write this with `min(len(self.classes_), self.n_components)`?

JPFrancoia replied:
Sure. You're right, it's more elegant.

jnothman replied:
Not merely more elegant, it's much easier to read the logic.

jnothman reviewed (changes requested):
otherwise lgtm

JPFrancoia committed:
Rewriting of a simple condition with min().

JPFrancoia committed:
Removed unecessary line.

jnothman commented:
but now you have test failures

JPFrancoia committed:
Addition of a condition to test if n_components is provided to the LDA
class. Used to determine the length of explained_variance_ratio_.

jnothman commented:
flake8 not happy

jnothman commented:
Your what's new should not be under 0.18.0

jnothman commented:
Otherwise, I'm okay with this, though I feel like there needs to be a bit more warning in the what's new about the subtle API change..

JPFrancoia committed:
Modification of whats_new to warn about the API change for
explained_variance_ratio.

JPFrancoia commented:
You're right. I updated the what's new. However, I don't know why the PR doesn't pass the flake8 test. Is it because a change I introduced ? It does not seem so.

jnothman commented:
It's because your PR was based on the 0.18.X branch. Generally we have all PRs to master and commits are cherry-picked back to version branches. I've changed the base, but you'll need to rebase locally.

JPFrancoia commented:
Ok, but there was a lot of changes regarding LDA that were in 0.18.X and not in master. I had to work from this version.

I never rebased with git. Should I run ```git rebase master 0.18.X``` locally ?

amueller commented:
There are no changes in LDA that are in 0.18.X but not in master.

amueller commented:
make sure you update master before rebasing on top of it. Something went wrong.

JPFrancoia commented:
This was too complicated for me. I opened a new PR, completely clean:

https://github.com/scikit-learn/scikit-learn/pull/7632

I'm closing this one.
